eth/db/journal.py

```diff
@@ -24,237 +26,258 @@ class DeletedEntry:
-        """
-        Setter for updating the *latest* changeset.
-        """
-        self.journal_data[self.latest_id] = value
-
-    def is_empty(self) -> bool:
-        return len(self.journal_data) == 0
+        # Interestingly, an attempt to cache this value caused a slowdown.
+        return first(reversed(self._journal_data.keys()))
 
-    def has_changeset(self, changeset_id: uuid.UUID) -> bool:
-        return changeset_id in self.journal_data
+    def has_changeset(self, changeset_id: JournalDBCheckpoint) -> bool:
+        # another option would be to enforce monotonically-increasing changeset ids, so we can do:
+        # checkpoint_idx = bisect_left(self._checkpoint_stack, changeset_id)
+        # (then validate against length and value at index)
+        return changeset_id in self._checkpoint_stack
 
-    def record_changeset(self, custom_changeset_id: uuid.UUID = None) -> uuid.UUID:
+    def record_changeset(
+            self,
+            custom_changeset_id: JournalDBCheckpoint = None) -> JournalDBCheckpoint:
         """
-        Creates a new changeset. Changesets are referenced by a random uuid4
+        Creates a new changeset. Changesets are referenced by a random int
         to prevent collisions between multiple changesets.
         """
         if custom_changeset_id is not None:
-            if custom_changeset_id in self.journal_data:
+            if custom_changeset_id in self._journal_data:
                 raise ValidationError(
                     "Tried to record with an existing changeset id: %r" % custom_changeset_id
                 )
             else:
                 changeset_id = custom_changeset_id
         else:
-            changeset_id = uuid.uuid4()
+            changeset_id = get_next_checkpoint_id()
 
-        self.journal_data[changeset_id] = {}
+        self._journal_data[changeset_id] = {}
+        self._checkpoint_stack.append(changeset_id)
         return changeset_id
 
-    def pop_changeset(self, changeset_id: uuid.UUID) -> Dict[bytes, Union[bytes, DeletedEntry]]:
-        """
-        Returns all changes from the given changeset.  This includes all of
-        the changes from any subsequent changeset, giving precedence to
-        later changesets.
-        """
-        if changeset_id not in self.journal_data:
-            raise KeyError(changeset_id, "Unknown changeset in JournalDB")
-
-        all_ids = tuple(self.journal_data.keys())
-        changeset_idx = all_ids.index(changeset_id)
-        changesets_to_pop = all_ids[changeset_idx:]
-        popped_clears = tuple(idx for idx in changesets_to_pop if idx in self._clears_at)
-        if popped_clears:
-            last_clear_idx = changesets_to_pop.index(popped_clears[-1])
-            changesets_to_drop = changesets_to_pop[:last_clear_idx]
-            changesets_to_merge = changesets_to_pop[last_clear_idx:]
+    def discard(self, through_checkpoint_id: JournalDBCheckpoint) -> None:
```

Comment:
To back it further, this method talks about `checkpoint_id` what is elsewhere referred to as `changeset_id`